Full-file syntax highlighting for diff pages

dfirebaugh opened the pull request:
Fix #33358, fix #21970

This adds a step in the `GitDiff` that does syntax highlighting for the entire file and then only references lines from that syntax highlighted code.  This allows things like multi-line comments to be syntax highlighted correctly.

![image (4)](https://github.com/user-attachments/assets/5d11b371-1940-41c1-9392-20c06955c3cc)
![image (3)](https://github.com/user-attachments/assets/99f5f61e-55d2-4e96-8096-0b7013640a8d)
![image (2)](https://github.com/user-attachments/assets/5ecfa633-b57d-4a6c-9508-d96e0368d027)
![image (1)](https://github.com/user-attachments/assets/e1e7ec59-ee13-40d4-91f4-0deb8e361e9e)
![image](https://github.com/user-attachments/assets/1f241a47-0640-4641-9707-3fcf8b328b2c)

wxiaoguang commented on services/gitdiff/highlightdiff_test.go:
```diff
@@ -20,11 +19,13 @@ func TestDiffWithHighlight(t *testing.T) {
 	)
 
-	expected := `		<span class="n">run</span><span class="o">(</span><span class="removed-code"><span class="k">&#39;</span><span class="o">&lt;</span><span class="o">&gt;</span><span class="k">&#39;</span></span><span class="o">)</span>`
+	expected := `		run(<span class="removed-code">'<>'</span>)
```
It can't be right. By your approach, there will be XSS. Something like `<script>xxx</script>` will be rendered to the user's page.

dfirebaugh replied:
Hey, thanks for reviewing.
I didn't change how html is encoded.

`<span class="removed-code">'<>'</span>` should render something like  `'<>'` to the page.

If I'm misunderstanding something, my apologies.


![Screenshot from 2025-03-02 00-33-03](https://github.com/user-attachments/assets/2227a0c3-fc43-449c-bac8-e7aa404ccd40)
![image](https://github.com/user-attachments/assets/7fd6e5a8-ced9-445b-8a43-ae871ac3aa5c)

wxiaoguang replied:
In your test result:

```
"run(<span class="removed-code">'<>'</span>)"
```

There are 3 `<` chars. The first one is `<span` (added by the render), the second one is the one in `'<>'`

The old logic does encode the second one to `&lt;`, which is necessary to render it to browser.

But in you case, the second one is still `<`, then it will be used as "HTML tag", but not rendered as "code content"

I do not see why it could be right.

dfirebaugh replied:
oh, it's because it's no longer calling `highlight.Code(filename, language, codeA)` here. So, it's not actually "rendering" the html here.

wxiaoguang replied:
It's not really related to "rendering" or not. Let's take a simpler example below:

You made this output:

```
`a <span class="removed-code"><</span> b`
```

What "kind" of the string it is?

* If it is a HTML string, then it is invalid, the `<` shouldn't appear in HTML content. It must be encoded to `&lt;`
* If it is not a HTML string, then `<span ...>` wouldn't be rendered as HTML but just be displayed as-is, actually it is not the case because this tag is for highlighting the code as removed.

I do not see it could be right.

wxiaoguang replied:
If you think your approach is right but the tests do not work for your approach, please design correct tests, and use the existing test data to verify that the logic overall is right.

wxiaoguang reviewed (dismissed):
~~HTML encoding should be correctly handled, for example: `<>`~~

See new comment below https://github.com/go-gitea/gitea/pull/33766#issuecomment-2692668999

wxiaoguang commented:
It also generated incorrect diff results (actually the "diff" content is corrupted)

Old:

![image](https://github.com/user-attachments/assets/2c246120-a148-4362-97cf-6683f3d4f25f)

Yours:

![image](https://github.com/user-attachments/assets/bc8ef35c-61e1-47a2-81f5-2929c10d6625)

wxiaoguang commented:
Thank you very much. I can understand most of changes now. The idea is feasible, while there are still some details:

1. Some of the old tests should be kept, for example: `diff.Text = "OC"`, I do not see why it should be removed. It is used to make sure the `recoverOneDiff` works correctly.
1. Some of the old tests do not fit the new code:
    * If a test doesn't fit (the case won't happen), feel free to remove it.
    * New tests should cover the new behavior since the behavior of `diffWithHighlight` has changed, it only accepts HTML content
1. Some variable types should be changed to match the new behavior, use `template.HTML`
1. Some edge cases, for example, what if a file is large (Mega-bytes)? It would bloat the Gitea's memory and make it OOM.
    * I think we need to have a limit, if a file is too large, fall back to the "block" diff but not the "full file" diff.

silverwind commented:
> I think we need to have a limit, if a file is too large, fall back to the "hunk block" diff but not the "full file" diff.

Sounds acceptable to have a certain byte limit above which to fall back to hunk-based diff. Maybe 5MB or so.

dfirebaugh commented:
Thank you,

since we fall back to hunk-based syntax highlighting (in some cases), all of the pre-existing tests should remain.  I've reverted those to the original tests with the original test data.

for the file limit, is it enough to check line count? or should it check actual MB?
I (perhaps naively) implemented a line count check based on a variable in settings.

I also implemented some of your other comments.

It may still need more.  I'm out of time for today, but will revisit if you have more thoughts/comments.

silverwind commented:
> for the file limit, is it enough to check line count? or should it check actual MB?
> I (perhaps naively) implemented a line count check based on a variable in settings.

Both are fine I guess, whatever is easier. For line count, maybe default to a value like 100k lines.

delvh commented:
Well… the line count doesn't prevent the issue, i.e. if someone uploads a list of large minified JSON objects with each object occupying one line.
Also, `os.Stat`ing is probably simpler than implementing an entire line-based checker.
So, my preference would be the size instead of line based check

wxiaoguang commented:
Managed to make some changes (haven't started the test case fixing):

1. Since `GetBlobLineCount` already reads the whole file content, we could simply reuse it to get "limited content"
    * only do entire-file highlighting when the file is not that large
2. Rewrite `GetComputedInlineDiffFor` to make it simpler and easier to maintain
3. Respect `DisableDiffHighlight`
4. Fallback to per-line-highlighting when the "limited content buffer" exceeds max limit.

Does it work as expected and/or look good to you?

wxiaoguang commented:
After reading "GitDiff" related code, I found that the legacy code is totally a mess ...... various patches were added again and again (for example: FileOnly vs GetPullDiffStats, GetPullRequestFiles vs ToCommit). I think we need to refactor the "diff" related code before adding more patches to it.

warasilapm commented:
> After reading "GitDiff" related code, I found that the legacy code is totally a mess ...... various patches were added again and again (for example: FileOnly vs GetPullDiffStats, GetPullRequestFiles vs ToCommit). I think we need to refactor the "diff" related code before adding more patches to it.

So... no syntax highlighting fix in the near future then? That'd be pretty disappointing.

wxiaoguang commented:
> So... no syntax highlighting fix in the near future then? That'd be pretty disappointing.

You could try to blame the related code and read the history, there are various patches to the GetDiff related function, some of them are "performance" related.

Undoubtedly, this change will significantly affects performance (see some FIXMEs I added, I do not know how much the affect would be at the moment)

It would be quite disappointing too when the performance regression occurs and affects a lot of users.

If it could be proven that there would be no performance regression, or if there would be someone who could fix the performance regression in first them when users complain, then we can merge it now. But if no, I think we'd better to be careful for this case.

wxiaoguang commented:
The first refactoring PR:  Refactor git diff #33810

warasilapm commented:
> It would be quite disappointing too when the performance regression occurs and affects a lot of users.

Hey, you know - fair enough.

wxiaoguang commented:
Most problems should have been resolved.

wxiaoguang commented on services/gitdiff/gitdiff.go:
```diff
@@ -283,116 +262,131 @@ type DiffInline struct {
+	}
+	if diffLineType != DiffLinePlain {
+		// it seems that Gitea doesn't need the line wrapper of Chroma, so do not add them back
+		// if the line wrappers are still needed in the future, it can be added back by "diffToHTML(hcd.lineWrapperTags. ...)"
```
```suggestion
		// if the line wrappers are still needed in the future, it can be added back by "diffLineWithHighlightWrapper(hcd.lineWrapperTags. ...)"
```

lunny commented:
When expend more codes, it result in 500

```
Render failed, failed to render template: repo/diff/blob_excerpt, error: template error: builtin(static):repo/diff/blob_excerpt:83:20 : executing "repo/diff/blob_excerpt" at <$.section.GetComputedInlineDiffFor>: error calling GetComputedInlineDiffFor: runtime error: invalid memory address or nil pointer dereference
----------------------------------------------------------------------
		{{$inlineDiff := $.section.GetComputedInlineDiffFor $line ctx.Locale}}
		                  ^
----------------------------------------------------------------------
```

wxiaoguang commented:
> When expend more codes, it result in 500

Fixed in b2dbc78bfd599dfc46e6faf322f132b1740cfea9 and added more tests

wxiaoguang commented:
Follow up fix:  Fix material icon & diff highlight #33844
